### PR TITLE
Reworked diffusion-dislocation creep model, added new test

### PR DESCRIPTION
**Pull request by bobmyhill**

### CODE REVIEW

File: tests/diffusion_dislocation.prm [hunk 1]
```diff
@@ -1,3 +1,7 @@
+# This somewhat specific test checks this the diffusion dislocation
```
Comment (jdannberg) [inline tests/diffusion_dislocation.prm]: checks the diffusion dislocation...

File: tests/diffusion_dislocation.prm [hunk 2]
```diff
@@ -1,3 +1,7 @@
...
+# rheology model (with Hirth and Kohlstedt (2004) parameters
+# for dry olivine
```
Comment (jdannberg) [inline tests/diffusion_dislocation.prm]: There is a ')' missing.

**Comment by bangerth** [comment]
You'll need to fix the indentation: http://www.math.clemson.edu/~heister/aspect-astyler-logs/cd2e54efc543671f01487b6758d2773f0758c272/log

There are also 3 failing tests: http://www.math.clemson.edu/~heister/aspect-logs/cd2e54efc543671f01487b6758d2773f0758c272/gcc The results differ in significant ways for one test (have you investigated why?) and there is an assertion for another one (the floating point exception) that would be worthwhile figuring out.

**Comment by bobmyhill** [comment]
Also corrected dislocation (and diffusion) creep laws to use only the part of the strain rate attributable to each process. This has caused the test outputs to change...

**Comment by bobmyhill** [comment]
Also corrected dislocation (and diffusion) creep laws to use only the part of the strain rate attributable to each process. This has caused the test outputs to change...

**Comment by bobmyhill** [comment]
Any suggestions, or is this ready to roll now? The bug fixes are pretty major, so it would be good to get the patches merged fairly soon.

File: include/aspect/geometry_model/chunk.h [hunk 3]
```diff
@@ -188,6 +188,9 @@ namespace aspect
             push_forward(const Point<dim> &chart_point) const;
         };
 
+        Point<dim> pull_back(const Point<dim>) const;
+        Point<dim> push_forward(const Point<dim>) const;
+
```
Comment (bangerth) [inline include/aspect/geometry_model/chunk.h]: These lines seem unnecessary for the current pull request, right? Can you remove them?

File: source/material_model/diffusion_dislocation.cc [hunk 4]
```diff
@@ -118,27 +135,76 @@ namespace aspect
...
     {
       // Viscosities
-      const double e2inv = second_invariant(strain_rate);
+      // In the first time step, strain rate is normally zero
+      // Sometimes it is also zero at other times.
+      // Let's set it to some very small number
+      double edot;
+      if (strain_rate.norm() == 0.0)
+        {
+          edot = 2.0*std::numeric_limits<double>::min();
```
Comment (bangerth) [inline source/material_model/diffusion_dislocation.cc]: Can you write this as
```
  const double edot = (strain_rate.norm() == 0.0
                                   ?
                                   2.0*std::numeric_limits<double>::min()
                                   :
                                   std::sqrt(abs(second_invariant(strain_rate))));
```
This way we get to make the variable `const`.

File: source/material_model/diffusion_dislocation.cc [hunk 5]
```diff
@@ -118,27 +135,76 @@ namespace aspect
...
-                                                      std::exp((activation_energies_dislocation[j]+
-                                                                activation_volumes_dislocation[j]*pressure)/(stress_exponents_dislocation[j])));
-
-          composition_viscosities[j] = std::min(std::max(std::pow((1.0/viscosity_diffusion + one_over_viscosity_dislocation), -1.0), min_visc), max_visc);
+          // Power law creep equation
+          // \eta = 0.5 * A_i^{-\frac{1}{n_i}} d^\frac{m}{n} e2inv_i^{\frac{1-n_i}{n_i}} \exp\left(\frac{E_i^* + PV_i^*}{n_iRT}\right)
+          // where i corresponds to diffusion or dislocation creep
+
+          double e2inv_diffusion = 0;
+          double e2inv_dislocation = e2inv;
```
Comment (bangerth) [inline source/material_model/diffusion_dislocation.cc]: why do you need these two variables? I don't see you modifying them?

These and several of the variables below can be made `const` as far as I can see.

Comment (bangerth) [inline source/material_model/diffusion_dislocation.cc]: Oh, I'm mistaken about the modification. Forget about the first sentence.

File: source/material_model/diffusion_dislocation.cc [hunk 6]
```diff
@@ -313,6 +379,12 @@ namespace aspect
           prm.declare_entry ("Reference viscosity", "1e22", Patterns::Double(0),
                              "Reference viscosity for nondimensionalization. Units $Pa s$");
 
+          // Viscosity iteration parameters
+          prm.declare_entry ("Relative viscosity tolerance", "1e-12", Patterns::Double(0),
+                             "Tolerance for correct diffusion/dislocation viscosity ratio.");
+          prm.declare_entry ("Viscosity maximum iterations", "40", Patterns::Integer(0),
+                             "Maximum number of iterations to find the correct"
+                             "diffusion/dislocation viscosity ratio.");
```
Comment (bangerth) [inline source/material_model/diffusion_dislocation.cc]: Need a space at the end of the second line here (after "correct")

File: source/material_model/diffusion_dislocation.cc [hunk 7]
```diff
@@ -336,43 +409,55 @@ namespace aspect
                              "viscosity at that point.  Select a weighted harmonic, arithmetic, "
                              "geometric, or maximum composition.");
           // Diffusion creep parameters
-          prm.declare_entry ("Prefactors for diffusion creep", "1.92e-11",
+          prm.declare_entry ("Prefactors for diffusion creep", "1.5e-15",
```
Comment (bangerth) [inline source/material_model/diffusion_dislocation.cc]: Is this backward compatible to the old model before this patch? If not, does it matter?

File: source/material_model/diffusion_dislocation.cc [hunk 8]
```diff
@@ -336,43 +409,55 @@ namespace aspect
...
+                             "List of grain size exponents, $m_diffusion$, for background mantle and compositional fields,"
+                             "for a total of N+1 values, where N is the number of compositional fields."
+                             "If only one values is given, then all use the same value.  Units: None");
+          prm.declare_entry ("Activation energies for diffusion creep", "375e3",
                              Patterns::List(Patterns::Double(0)),
                              "List of activation energies, $E_a$, for background mantle and compositional fields,"
                              "for a total of N+1 values, where N is the number of compositional fields."
                              "If only one values is given, then all use the same value.  Units: $J / mol$");
-          prm.declare_entry ("Activation volumes for diffusion creep", "6.4e-6",
+          prm.declare_entry ("Activation volumes for diffusion creep", "6e-6",
```
Comment (bangerth) [inline source/material_model/diffusion_dislocation.cc]: same here and below.

File: source/material_model/diffusion_dislocation.cc (hunk 6 above)
Comment (bangerth) [inline source/material_model/diffusion_dislocation.cc]: The same is true for several of the multiline documentation strings below.

File: source/simulator/helper_functions.cc [hunk 9]
```diff
@@ -1120,7 +1120,10 @@ namespace aspect
           for (unsigned int q=0; q<output.size(); ++q)
             {
               Tensor<1,dim> g = gravity_->gravity_vector(in.position[q]);
-              output[q] = std::fabs(std::min(0.0, g*velocity_values[q]/g.norm()))*year_in_seconds;
+              if (g.norm() < 2.0*std::numeric_limits<double>::min())
+                output[q] = 0.0;
+              else
+                output[q] = std::fabs(std::min(0.0, g*velocity_values[q]/g.norm()))*year_in_seconds;
```
Comment (bangerth) [inline source/simulator/helper_functions.cc]: why is this necessary? I'd rather you handle this in the place where you consume the number computed here, rather than here.

File: source/material_model/diffusion_dislocation.cc [hunk 10]
```diff
@@ -518,19 +541,19 @@ namespace aspect
...
+                                   " $\\dot{\\varepsilon}$ is the second invariant of the strain rate tensor,"
+                                   " $R$ is the gas constant, $T$ is temperature, and $P$ is pressure."
+                                   " $A_i$ are prefactors, $n_i$ and $m_i$ are stress and grain size exponents"
                                    " $E_i$ are the activation energies, $V_i$ are the activation volumes,"
-                                   " $n_dislocation$ are stress exponents for dislocation creep,"
-                                   " $\\rho_m$ is the mantle density, $R$ is the gas constant,"
-                                   " $T$ is temperature, and $P$ is pressure."
                                    " \n\n"
                                    " The value for the components of this formula and additional"
                                    " parameters are read from the parameter file in subsection"
```
Comment (bangerth) [inline source/material_model/diffusion_dislocation.cc]: The documentation does not talk about any iteration, but you have parameters that seem to imply that you iterate some sort of nonlinearity. Can you add a paragraph that explains this procedure?

File: tests/diffusion_dislocation_fixed_strain_rate.prm [hunk 11]
```diff
@@ -0,0 +1,101 @@
+# This model has a box geometry where simple shear is induced by constant velocity
+# along the bottom and top boundaries. Diffusion dislocation creep laws corresponding
+# to the model of Hirth and Kohlstedt for dry olivine are used, and the viscosity is
+# calculated at the first time step for a constant temperature of 1600 K.
+# The analytical solution for the viscosity with the given parameters is 5.81735e17 Pas.
+# Note that the strain rate output by aspect is the *norm*, but the model itself uses
+# the square root of the second invariant. For reference, edot_norm = sqrt(2*edot_ii).
+# The depth average output contains viscosity output, and so provides the required test.
```
Comment (bangerth) [inline tests/diffusion_dislocation_fixed_strain_rate.prm]: nice explanation, thank you!

**Comment by bangerth** [comment]
I have only a few comments. Please address them, squash the various commits into one (or, if you want to, separate the tests into a second commit) and re-upload. Thanks for the work!

File: source/material_model/diffusion_dislocation.cc (hunk 7 above)
Comment (bobmyhill) [inline source/material_model/diffusion_dislocation.cc]: The old model before the patch gave the incorrect answer (for the same reasons as the Morency Doin model), so I took the opportunity to change the parameters to something more relevant.

File: source/simulator/helper_functions.cc (hunk 9 above)
Comment (bobmyhill) [inline source/simulator/helper_functions.cc]: The original line produced a divide by zero exception when using the depth average postprocessor if g in a parameter file was equal to zero. This is what we do in the fixed strain rate test, but I'm sure there are other prm files out there with g=0 which would also fail if we tried to use the depth average postprocessor. 

What do you think?

File: tests/diffusion_dislocation_fixed_strain_rate.prm (hunk 11 above)
Comment (bobmyhill) [inline tests/diffusion_dislocation_fixed_strain_rate.prm]: Welcome :)

**Comment by bobmyhill** [comment]
Ok, I've made the changes that didn't need any more discussion. The one thing which might still need to be changed is the bug fix in helper_functions.cc, but I think that it's probably necessary, given the new divide by zero exceptions.

File: source/material_model/diffusion_dislocation.cc (hunk 7 above)
Comment (bangerth) [inline source/material_model/diffusion_dislocation.cc]: Ah, fair enough.

File: source/simulator/helper_functions.cc (hunk 9 above)
Comment (bangerth) [inline source/simulator/helper_functions.cc]: Ah, I see. It's unrelated to the rest of the patch then. How about something like
```
  const Tensor<1,dim> g = ...;
  const Tensor<1,dim> vertical = (g.norm() > 0 ? g/g.norm() : Tensor<1,dim>());
  output[q] = std::fabs(std::min(0.0, velocity_values[q]*vertical))*year_in_seconds;
```
? 

This kind of thing is best handled through a separate pull request, rather than mixing it into a bigger issue. If you'd like to, I can create such a separate request, or I can leave it to you.

Comment (bangerth) [inline source/simulator/helper_functions.cc]: On second thought: This was so easy that it is now #529. 

Just remove the changes you made here then, and the fix will come in through the separate pull request.

**Comment by bangerth** [comment]
Thanks, @bobmyhill ! 

1/ Please just take one the changes in the one place in `helper_functions.cc` and we get the corresponding functionality from #529.

2/ I did not realize this in the original review apparently, but right now the parameters are all put into a section declared via
```
     prm.enter_subsection ("DiffusionDislocation");
```
We have usually used "Separate words, first one capitalized" style for subsections. This would suggest
```
     prm.enter_subsection ("Diffusion dislocation");
```
Was this model already in the last release? I think it wasn't, so we could still change this. Would you be willing to?

3/ There are currently 3 failing tests (see http://www.math.clemson.edu/~heister/aspect-logs/c06a4e30ff8e026daf12e2b75fa41754a908e801/gcc):
```
	 75 - diffusion_dislocation (Failed)
	 76 - diffusion_dislocation_fixed_strain_rate (Failed)
	108 - material_model_dependencies_diffusion_dislocation (Failed)
```
What output do they produce for you? Can you just take the output from your machine and copy it over the existing output, to see whether the tester likes it better?

**Comment by bobmyhill** [comment]
Ok, I've made the suggested changes, and changed the iterative diffusion:dislocation ratio loop to use the Newton method - this should converge faster and be more readily understandable.

**Comment by bangerth** [comment]
Great. There are currently 3 failiing tests, of which `output-material_model_dependencies_diffusion_dislocation` produces a floating point exception (most likely a division by zero). Can you take a look where this is from?

Other than that, the remaining two tests seem to be harmless small deviations that Timo can address in a merge. The one above I'd like to understand, though.

**Comment by bobmyhill** [comment]
The model doesn't work if the temperature is too low - the exponential term causes the stress prefactor to approach zero, and the inverse is used on line 174 of source/material_model/diffusion_dislocation.cc. I've changed the test so that it doesn't use very low numbers. 

Anyone using this model should realise that as temperature approaches 0 K, things will go seriously wrong. I'd be ok with raising an exception if the temperature in the model gets seriously low, but would rather not add "if stress prefactor tiny, then make bigger", as this would encourage users to implement something both physically unreasonable and likely to break.

File: tests/material_model_dependencies_diffusion_dislocation/screen-output [hunk 12]
```diff
@@ -2,44 +2,36 @@
 -- This is ASPECT, the Advanced Solver for Problems in Earth's ConvecTion.
 --     . version 1.4.pre
 --     . running in DEBUG mode
---     . running with 1 MPI process
+--     . running with 4 MPI processes
 --     . using Trilinos
 -----------------------------------------------------------------------------
 
-Loading shared library <./libmaterial_model_dependencies_diffusion_dislocation.so>
-
```
Comment (bangerth) [inline tests/material_model_dependencies_diffusion_dislocation/screen-output]: you'll need to run this test via ctest, or it will not load the shared lib and not create the relevant output below (which also disappeared from your test output). Can you run it again and update the screen-output?

**Comment by bobmyhill** [comment]
Ok, hopefully this should be better. Sorry about that!

**Comment by bangerth** [comment]
Looks good.

@tjhei : Would you mind updating the tests and merging this?

**Comment by gassmoeller** [comment]
Little reminder for @tjhei to update the tests ;-). I can merge it afterwards.

**Comment by bangerth** [comment]
@tjhei seems to be MIA. @bobmyhill -- can you rebase this to the current state and re-upload the branch? I think that should trigger another run of the tester which I believe now produces a reasonable diff of the tests that one could then use to manually fix the test output -- at which point I (or @gassmoeller) could merge myself. (Same for #524 if you want.)

**Comment by gassmoeller** [comment]
Ok, it seems the tester is happy now. However, could you please go through the parse_parameters function and add spaces to the assignments (`x = a`, not `x=a`). Just to have a consistent style, and I promise to merge afterwards ;-).

**Comment by bobmyhill** [comment]
Promises, promises ... 
:P
